### ISSUE TITLE
Refactoring

### ISSUE THREAD

**Issue by mottosso** [pyblish/pyblish-qml#71]
# Goal

Simplify logic, separate concerns and improve robustness.

# Implementation

This thread acts as a summary of changes made during the past few days of refactoring.

- Lock-free processing
- Implementation of Statecharts
- Separation of concerns
- Tests

### Lock-free processing

The processing algorithm implemented prior to this refactoring were processing in such a way that required locking and releasing of a thread while drawing and evaluating the current state.

```python
# psuedo-code
while True:
    for plugin in plugins:
        # Is this a plug-in we care about processing?
        if not plugin.is_toggled:
            continue

        # Given the current state of processing, should we continue?
        if not self.test(plugin):
             break

        result = plugin.process()

        lock()
        # Draw to screen from main thread
        self.draw(result)
        unlock()
```

The `test()` evaluates whether or not to continue processing - e.g. a validator may have failed - whereas `draw()` must occur within the main thread.

The refactored implementation however both tests and draws from the main thread, whilst initiating a new process only once both have finished.

```python
# psuedo-code
def next(result=None):
    if result:
       self.draw(result)
   
    # In place of the loop, a Python generator handles
    # the logic of figuring out which plug-in to process next.
    next = self.compute_next()

    if self.test(next):
        # Run in separate thread, return result back to `next()`
        threaded(self.process, args=[item], callback=self.next)
```

Instead of a loop, this algorithm is recursive and greatly simplifies the logic involved. Now we are free to hold off processing, without keeping a live thread on hold and without locking.

With this approach it is also possible to decouple drawing from processing further, so as to perform both asynchronously, thus speeding up the overall time taken to complete due to one not having to wait for the other.

Additionally, due to having refactored both testing and computation of the next item, we can pass different tests and generators to the same algorithm yielding different results.

```python
def next(plugin, test, generator):
      item = generator.next()
      if test(item):
          threaded(self.process, args=[item], callback=self.next)
```

This facilitates use in different functions, along with testing as we can now pass it arbitrary generators and tests to make sure it behaves the way we expect it to.

### Harel Statecharts

In response to the increasing amount of signals and events taking place in both the Controller and Presentation layer, I've gone replaced the previous ad-hoc handling with statecharts. 

- http://en.wikipedia.org/wiki/State_diagram#Harel_statechart

Having researched about it during the past few weeks, it seemed a good fit for the event-based nature that is Qt.

As an example, the following states were introduced in the Controller.

- `hidden`
- `visible`
- `ready`
- `publishing`
- `finished`
- `repairing`
- `initialising`
- `stopping`
- `saving`
- `stopped`
- `dirty`
- `clean`

Some of which are parallel and in which some states allow for certain actions to take place whereas others do not - such as being able to publish from `ready`, but not from `resetting`. The result is more predictability and a convergence amongst signals and events taking place throughout the GUI.

Furthermore, the Presentation layer is able to take advantage of these states to provide the appropriate graphical items corresponding to each state - such as providing a `Publish` button when 
`ready`, and `Stop` button when in `publishing`.

This prevents the user interface from asking the Controller to do things it isn't in a state to perform, and vice versa, in addition to maintaining a top-level representation of what state the overall application is in; simplifying testing and debugging.

### Separation of concerns

Previously, business logic such as *"is instance A compatible with plugin B?"* was implemented both in Pyblish Core and Pyblish QML. This was due to the fact that Pyblish QML is communicating with a host, across processes, via Endpoint and thus not having direct access to Pyblish Core.

Implementing logic twice meant both a duplication in effort but also a disparity between behaviours - for example determining whether a given plug-in is compatible with a given plug-in recently yielded different results in the GUI from how it did in Core.

### Tests

From a high-level perspective, the application consists of three parts.

- Model
- View
- Controller

Our model is our host, as provided by Endpoint, and communication to it is made through TCP via standard GET, POST and PUT requests. This part is tested thoroughly and individually in the Endpoint repository.

The view is our QML layer and is tested using QML testing tools, which are all written in Javascript and performed via the Qt test-runner.

- http://doc.qt.io/qt-5/qtquick-qtquicktest.html

Finally, the controller is where the meat of the application is held and represents a `QtCore.QObject` subclass implementing logic and behaviour. Once the user hits `Publish`, a signal is sent from QML to Python, our controller, which triggers the publishing event loop. The controller is tested individually as well using standard Python unittesting with help from the QtTest module.

Tests for the controller were implemented in a data-driven fashion; in which the same test operates on multiple sets of data. The data in this case being various combinations of plug-ins and instances. This allows for tests to stay general, and for a large amount of *scenarios* to be tested with little effort.

**Comment by mottosso** [comment]
A note on the current data-driven approach.

We're using pre-formatted dictionaries that conform to the Schemas defined by Endpoint, as opposed to actual plug-ins written using `pyblish.api`.

The reason for this is to avoid too much re-implementation of the formatting mechanism provided by Endpoint, and instead manually format only data that is relevant to the tests.

Case in point, detecting whether a plug-in has any compatible instances and vice versa, and whether a plug-in can process a context or instance.

By hand-writing this data, we avoid the risk of wrongly serialising and updating the mechanism, at the cost of re-writing the test data. Both are tedious in case of change, but I don't see any other way.

**Using test-data**

```python
import os
import json


Plugin = "plugin"
Instance = "instance"

_cache = {}
module_dir = os.path.dirname(__file__)


def load(type, name):
    print "data.load(%s, %s)" % (type, name)
    fname = "%s_%s.json" % (type, name)

    if fname not in _cache:
        path = os.path.join(module_dir, "data", fname)
        with open(path, "r") as f:
            _cache[fname] = json.load(f)

    data = _cache[fname]
    data["name"] = name
    return data


def dump(type, name, indent=4):
    return json.dumps(load(type, name), indent=indent, sort_keys=True)
```

Data looks like this.

```python
# instance_Instance1
{
    "__fails__": false,

    "data": {
        "family": "family1"
    }
}

# plugin_Validator1
{
    "__fails__": false,

    "data": {
        "families": ["family1"],
        "order": 1,
        "canProcessInstance": true
    }
}
```

**Using actual instances**

```python
import os

# Dependencies
import pyblish.api

import util

Plugin = "plugin"
Instance = "instance"

_cache = {}
module_dir = os.path.dirname(__file__)


def load(type, name):
    if type == "plugin":
        return plugins[name]
    if type == "instance":
        return instances[name]
    raise ValueError("Could not load %s, %s" % (type, name))


@pyblish.api.log
class Selector1(pyblish.api.Selector):
    hosts = ["*"]

    def process_context(self, context):
        for instance in instances:
            instance = context.create_instance(name=instance["name"])
            for key, value in instance["data"].items():
                instance.set_data(key, value)


@pyblish.api.log
class Validator1(pyblish.api.Validator):
    families = ["family1"]
    hosts = ["*"]

    def process_instance(self, instance):
        pass


@pyblish.api.log
class ValidateIncompatible(pyblish.api.Validator):
    families = ["familyIncompatible"]
    hosts = ["*"]


@pyblish.api.log
class Extractor1(pyblish.api.Extractor):
    families = ["family1"]
    hosts = ["*"]
    optional = True

    def process_instance(self, instance):
        pass


@pyblish.api.log
class Conformer1(pyblish.api.Conformer):
    families = ["family1"]
    optional = True

    def process_instance(self, instance):
        pass


plugins = util.ItemList(key="__name__")
plugins[:] = [
    Selector1,
    Validator1,
    ValidateIncompatible,
    Extractor1,
    Conformer1
]

instances = util.ItemList(key="name")
instances[:] = [
    {
        "name": "Instance1",
        "data": {
            "family": "family1"
        }
    },
    {
        "name": "Instance2",
        "data": {
            "family": "family1"
        }
    }
]
```
